// react-clouds/dist/index.esm.js
import React, { useState, useRef, useEffect } from 'react';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var Clouds = function (_a) {
    var _b = _a.size, size = _b === void 0 ? 100 : _b, _c = _a.color, color = _c === void 0 ? '#fff0f3' : _c, _d = _a.bottomOffset, bottomOffset = _d === void 0 ? 0 : _d, _e = _a.xMinMultiplier, xMinMultiplier = _e === void 0 ? 0.45 : _e, _f = _a.xMaxMultiplier, xMaxMultiplier = _f === void 0 ? 0.6 : _f, _g = _a.yMinMultiplier, yMinMultiplier = _g === void 0 ? 0.05 : _g, _h = _a.yMaxMultiplier, yMaxMultiplier = _h === void 0 ? 0.25 : _h, _j = _a.initialXOffset, initialXOffset = _j === void 0 ? 0.45 : _j, _k = _a.className, className = _k === void 0 ? '' : _k, _l = _a.style, style = _l === void 0 ? {} : _l;
    var _m = useState(''), boxShadow = _m[0], setBoxShadow = _m[1];
    var cloudRef = useRef(null);
    var generateShadows = function () {
        var _a;
        if (!cloudRef.current)
            return;
        var containerWidth = ((_a = cloudRef.current.parentElement) === null || _a === void 0 ? void 0 : _a.offsetWidth) || window.innerWidth;
        var shadows = [];
        var currentX = size * initialXOffset;
        while (currentX < containerWidth + size) {
            // Generate random Y offset within the defined range
            var yOffset = Math.floor(Math.random() * (size * (yMaxMultiplier - yMinMultiplier)) +
                (size * yMinMultiplier));
            shadows.push("".concat(currentX, "px ").concat(yOffset, "px ").concat(color));
            // Calculate next X position with random increment
            var xIncrement = size * (xMinMultiplier + Math.random() * (xMaxMultiplier - xMinMultiplier));
            currentX += xIncrement;
        }
        setBoxShadow(shadows.join(', '));
    };
    useEffect(function () {
        generateShadows();
        // Handle window resize
        var handleResize = function () {
            generateShadows();
        };
        window.addEventListener('resize', handleResize);
        return function () { return window.removeEventListener('resize', handleResize); };
    }, [
        size,
        color,
        xMinMultiplier,
        xMaxMultiplier,
        yMinMultiplier,
        yMaxMultiplier,
        initialXOffset
    ]);
    return (React.createElement("div", { ref: cloudRef, className: className, style: __assign({ position: 'absolute', width: "".concat(size, "px"), height: "".concat(size, "px"), borderRadius: '50%', backgroundColor: color, zIndex: 5, bottom: "".concat(-size * 0.1 + bottomOffset, "px"), left: "".concat(-size * 0.2, "px"), boxShadow: boxShadow }, style) }));
};

var CloudLayers = function (_a) {
    var _b = _a.className, className = _b === void 0 ? '' : _b, _c = _a.style, style = _c === void 0 ? {} : _c, _d = _a.backgroundColor, backgroundColor = _d === void 0 ? '#ffa6b9' : _d, _e = _a.height, height = _e === void 0 ? '24rem' : _e, _f = _a.shadowOpacity, shadowOpacity = _f === void 0 ? 0.1 : _f, _g = _a.layers, layers = _g === void 0 ? [
        {
            size: 120,
            color: '#ffdbde',
            bottomOffset: 160,
            randomization: {
                xMin: 0.5,
                xMax: 0.7,
                yMin: 0.1,
                yMax: 0.3,
                initialX: 0.4
            }
        },
        {
            size: 110,
            color: '#ffe4e9',
            bottomOffset: 80,
            randomization: {
                xMin: 0.45,
                xMax: 0.65,
                yMin: 0.05,
                yMax: 0.25,
                initialX: 0.45
            }
        },
        {
            size: 100,
            color: '#ffffff',
            bottomOffset: 0,
            randomization: {
                xMin: 0.4,
                xMax: 0.6,
                yMin: 0.03,
                yMax: 0.2,
                initialX: 0.5
            }
        }
    ] : _g;
    return (React.createElement("div", { className: "relative ".concat(className), style: __assign({ backgroundColor: backgroundColor, height: height, width: '100%', overflow: 'hidden' }, style) }, layers.map(function (layer, index) {
        var _a, _b, _c, _d, _e;
        return (React.createElement("div", { key: index, style: {
                position: 'absolute',
                width: '100%',
                height: '100%',
                filter: "drop-shadow(0 2px 3px rgba(0, 0, 0, ".concat(shadowOpacity * (1 - index * 0.2), "))")
            } },
            React.createElement(Clouds, { size: layer.size, color: layer.color, bottomOffset: layer.bottomOffset, xMinMultiplier: (_a = layer.randomization) === null || _a === void 0 ? void 0 : _a.xMin, xMaxMultiplier: (_b = layer.randomization) === null || _b === void 0 ? void 0 : _b.xMax, yMinMultiplier: (_c = layer.randomization) === null || _c === void 0 ? void 0 : _c.yMin, yMaxMultiplier: (_d = layer.randomization) === null || _d === void 0 ? void 0 : _d.yMax, initialXOffset: (_e = layer.randomization) === null || _e === void 0 ? void 0 : _e.initialX })));
    })));
};

export { CloudLayers, Clouds };
